clap_derive/tests/ui/external_subcommand_wrong_type.stderr

```diff
@@ -0,0 +1,8 @@
+error[E0308]: mismatched types
+  --> $DIR/external_subcommand_wrong_type.rs:13:15
+   |
+13 |     Other(Vec<CString>),
+   |               ^^^^^^^ expected struct `std::ffi::CString`, found struct `std::ffi::OsString`
```

Comment:
Reverse order?

Reply:
We have no control over the error message here, only the span can be adjusted. I'd [been trying](https://github.com/TeXitoi/structopt/pull/314#discussion_r365519669) to do something SFINAE-like but didn't succeed.

Reply:
I thought we should getting aborted with [this](https://github.com/clap-rs/clap/pull/1883/files#diff-f2fd0b2248dc9337423a3cc26e1b26a9R162) error message.

Reply:
@CreepySkeleton Am I mistaken here?

Reply:
Not sure what message you was referring to. To make it clear: you want it to look like this, right?:
```
--> $DIR/external_subcommand_wrong_type.rs:13:15
   |
13 |     Other(Vec<CString>),
   |               ^^^^^^^ expected struct `std::ffi::OsString`, found struct `std::ffi::CString`
```

Reply:
I am saying that this message shouldn't happen at all because we should be aborting much earlier.

Reply:
This one is on purpose. It checks that the type here is either `Vec<String>` (and is detected) or it defaults to `OsString` (which is what happening here and causes the error). It must happen.

Reply:
It gives a bad error message which is not really informative. That's why I was saying we should probably abort with a better informative message instead of trying to get it to work.

Reply:
In order to abort, we'd have to detect that the field is neither `String` nor `OsString` somehow. But how do we detect it? The algorithm is pretty dumb: check for `Vec<String>`, if yes, generate code for string, otherwise cross fingers and generate code for `OsString`. If it happens to be _not_ `OsString`, you'll see the error above which kind of mentions `OsString` and this should be enough for it to be useful.

Reply:
I am confused, why generate for `OsString`, why not also check for `OsString`?

Reply:
Because we would have to special case it, like we do with `String`, and I'm really not cool with special casing types outside of prelude. I'm not OK because people tent to refer it not only as `OsString` but also by `ffi::OsString` (I saw it a lot) and various aliases.

Reply:
Okay, please add the requested test cases and we can merge this. But I would like to take a look later.

Reply:
Theoretically speaking, I think you may be able to make a sealed `StringOrOsString` trait, implement it exclusively for `String` and `OsString` and then generated code would be able to use it for converting from  `&OsStr` and fill the `Vec`.